File: server/src/remotestdio.rb
#  typed: true
  
require 'net/http'
require 'uri'
require 'json'
require 'delegate'
require 'logger'

require 'sorbet-runtime'
require 'websocket-client-simple'

extend T::Sig

LOGGER = Logger.new(STDOUT)
set_level = ENV['LOG_LEVEL'].to_s.to_sym
LOGGER.level = ( (Logger.constants.include?(set_level) && Logger.const_get(set_level)) || Logger::INFO )


class WhineThread < Thread
  extend T::Sig
  sig {params(blk: T.proc.void).void}
  def initialize(&blk)
    super() {
      begin
        blk.call
      rescue StandardError => ex
        STDERR.puts(ex.message, ex.backtrace&.join("\n"))
      end
    }
  end
end

class WSIN
  extend T::Sig
  attr_accessor :got_msg

  sig {params(host: String, client_id: String).void}
  def initialize(host, client_id)
    @client_id = client_id
    @host = host
    @msg = ''
    @got_msg = T.let(false, T.any(T::Boolean, Integer))
    @connection = WhineThread.new {
      # ws simple's self in ws.on block is not WSIN!
      msg = @msg
      got_msg = @got_msg
      client_id = @client_id
      q = Queue.new

      loop do
        ws = WebSocket::Client::Simple.connect("ws://#{host}/websocket/back")

        ws&.on :open do |event|
          LOGGER.debug("WSIN opend. send ack #{client_id.inspect}")
          ws.send(client_id)
        end

        ws&.on :message do |event|
          LOGGER.debug("WSIN msg #{event.data.inspect}")
          data = JSON.parse(event.data, symbolize_names: true)
          status = T.let(data[:status].to_i, Integer)
          msg, got_msg = if status == 200 then
            tmp = T.let(data[:data].to_s, String)
            # [msg, got_msg]
            [tmp, true]
          else
            ['', status]
          end
          q << got_msg
          ws.close()
        end

        ws&.on :close do |event|
          LOGGER.debug("WSIN :close #{event.code} #{event.reason}")
          if got_msg then
            ws = nil
            EM.stop
            LOGGER.error("WSIN refused. already stdin opened?") if got_msg == 403
          else
            LOGGER.warn("WSIN closed before msg. Retry after 10s")
            sleep 10
          end
        end

        q.pop # wait
        break if got_msg
      end # loop

      @got_msg = got_msg
      @msg     = msg
    }
  end

  sig{ params(a: String).returns(String)}
  def gets(*a)
    @connection.join
    return @msg
  end
end

class RemoteSTDIO
  class << self
    extend T::Sig

    sig {params(host: String, id: String).void}
    def init(host, id)
      @host = host # hostname:port
      @url = url = URI.parse("http://#{host}/stdout")
      # HTTPリクエストの作成
      @http = Net::HTTP.new(url.host, url.port)
      @http.use_ssl = (url.scheme == 'https')

      @client_id = id
      @mutex = Mutex.new

      $stderr = FakeSTDERR.new($stderr)
    end

    sig {params(msg: String, err: T::Boolean).returns(T.untyped)}
    def write(msg, err: false)
      return if @client_id.nil?

      payload = { message: msg, client_id: @client_id, err: err }.to_json
      request = Net::HTTP::Post.new(@url.path, { 'Content-Type' => 'application/json' })
      request.body = payload

      # リクエスト送信
      return response = @http.request(request)

      # レスポンス表示
      # puts "Response Code: #{response.code}"
      # puts "Response Body: #{response.body}"
    end

    sig {params(a: T.untyped, err: T::Boolean, noremote: T::Boolean).returns(NilClass)}
    def puts(*a, err:false, noremote: false)
      self.write(a.map(&:to_s).join("\n") + "\n", err: err) if !noremote
      return nil
    end

    sig {params(a: T.untyped, err: T::Boolean, noremote: T::Boolean).returns(NilClass)}
    def print(*a, err:false, noremote: false)
      self.write(a.map(&:to_s).join, err: err) if !noremote
      return nil
    end

    sig {params(a: T.untyped, err: T::Boolean, noremote: T::Boolean).returns(NilClass)}
    def p(*a, err:false, noremote: false)
      self.write(a.map(&:inspect).join, err: err) if !noremote
      return nil
    end

    sig {params(retval: T.nilable(String), threads: T::Hash[Symbol, T.nilable(WhineThread)], counter_key: Symbol, trial: Integer).returns(T.nilable(String))}
    def handle_concurrency(retval, threads, counter_key, trial: 3)
      @mutex.synchronize do
        counter = catch :counter do
          trial.times {
            canbe_couter = threads[counter_key]
            throw :counter, canbe_couter unless canbe_couter.nil?
            sleep 0.5
          }
          LOGGER.error("Failed to get #{counter_key} thead for gets")
          nil
        end
        return nil if counter.nil?

        if counter&.status then
          LOGGER.debug("Will kill #{counter_key} stdin")
          counter.kill
        end
      end
      return retval
    end

    sig {params(a: String, noremote: T::Boolean, kw: String).returns(String)}
    def gets(*a, noremote: false, **kw)
      return STDIN.gets(*T.unsafe(a), **kw) if @client_id.nil? || noremote

      threads = T.let({}, T::Hash[Symbol, WhineThread])
      wsin = WSIN.new(@host, @client_id)
      @retval = T.let(nil, T.nilable(String))

      threads[:STDIN] = WhineThread.new {
        retval = T.let(STDIN.gets(*T.unsafe(a)), T.nilable(String))
        LOGGER.debug("STDIN got #{retval.inspect}")
        if !retval.nil? then # if nil, close STDIN thread and wait wsin
          wsin.got_msg = true # break wsin gets loop to avoid loop even after ws.close
          @retval = handle_concurrency(retval, threads, :wsin) 
        end
      }
      threads[:wsin] = WhineThread.new {
        retval = wsin.gets(*T.unsafe(a))
        LOGGER.debug("wsin got #{retval.inspect}")
        @retval = handle_concurrency(retval, threads, :STDIN) 
      }

      threads.each{|_, t| t.join }
      return @retval.to_s
    end

  end
end

class FakeSTDERR < SimpleDelegator
  def puts(*a)
    STDERR.puts(*T.unsafe(a))
    RemoteSTDIO.puts(*T.unsafe(a), err: true)
  end

  def print(*a)
    STDERR.print(*T.unsafe(a))
    RemoteSTDIO.print(*T.unsafe(a), err: true)
  end
end

module Kernel
  %i[puts print gets].each{|method_name|
    orig_name = :"#{method_name}_orig"
    alias_method(orig_name, :"#{method_name}")
    define_method(method_name){|*a, **kw|

      #$stdout.puts("Orig: #{method_name}")
      tmp = [orig_name, *a]
      send(*T.unsafe(tmp)) if method_name != :gets
      tmp = [method_name, *a]
      RemoteSTDIO.send(*T.unsafe(tmp), **kw)
    }
  }
end

module RemoteSTDIOUtils
  include Kernel
  extend self
  attr_accessor :no_remotestdio
  @no_remotestdio = false

  def init_by_envvar()
    host = ENV['HOST'].to_s
    client_id = ENV['CID'].to_s
    if !host.empty? && !client_id.empty? then
      RemoteSTDIO.init(host, client_id)
    else
      puts "WARN puts without remotestdio"
    end
  end

  def safe_gets()
    if RemoteSTDIOUtils.no_remotestdio then
      return STDIN.gets
    else
      return gets
    end
  end
end


# class CIO
#   def write(msg)
#     STDOUT.puts("orig: #{msg.inspect}")
#     STDOUT.write(msg)
#   end
# 
#   def read(msg)
#     STDOUT.print(msg)
#     STDIN.read(msg)
#   end
#   def readline(msg)
#     STDOUT.print(msg)
#     STDIN.readline(msg)
#   end
#   def gets(*a)
#     STDOUT.print('hi',*a)
#     STDIN.gets(*a)
#   end
# end
# 
# $stdin = CIO.new

if __FILE__ == $0
  RemoteSTDIO.init(T.must(ENV['HOST']), T.must(ENV['CID']))
  if ARGV.delete_at(0)&.downcase&.include?(?a) then
    print(
"""## Hello world
**At #{Time.now.iso8601}**
~~Test text~~
[line-through]#Test text#

* [.red]#赤くする#
* [.green]#緑にする#

[.navy]#Input# >>""")
  else
    print(
"""## Hello world
**At #{Time.now.iso8601}**
~~Test text~~
````
## This is
```bash
echo raw code ${HOME}
```
````

```bash
$ ls ${HOME}
```

Input >>""")
  end
  val = gets()
  $stderr.puts("Got #{val}")
  p(val)
end

File: server/src/remotestdio.py
# Thank you Gemini 2.5 Pro

import sys
import os
import threading
import json
import time
from urllib.parse import urlparse
import requests
import websocket  # `pip install websocket-client requests` が必要です
import builtins

class WhineThread(threading.Thread):
    """
    Ruby版のWhineThreadと同様に、スレッド内で発生した例外を
    標準エラー出力に表示するスレッドクラス。
    """
    def __init__(self, group=None, target=None, name=None, args=(), kwargs=None, *, daemon=None):
        super().__init__(group=group, target=target, name=name, args=args, kwargs=kwargs, daemon=daemon)
        self._target = target
        self._args = args
        self._kwargs = kwargs if kwargs is not None else {}

    def run(self):
        """スレッドの実行ロジックをラップし、例外を捕捉します。"""
        try:
            if self._target:
                self._target(*self._args, **self._kwargs)
        except Exception as e:
            # スレッド内で発生したエラーを標準エラー出力に書き出す
            sys.stderr.write(f"Error in thread {self.name}: {e}\n")
            import traceback
            sys.stderr.write(traceback.format_exc())

class WSIN:
    """
    WebSocketサーバーに接続し、メッセージを受信するためのクラス。
    RubyのWSINクラスのロジックを再現します。
    """
    def __init__(self, host, client_id):
        self.host = host
        self.client_id = client_id
        self.got_msg = False
        self.msg = ''

    def gets(self):
        """
        WebSocketサーバーからメッセージを受信するまでブロックします。
        接続が切れた場合は10秒待ってから再試行します。
        """
        while not self.got_msg:
            try:
                ws_url = f"ws://{self.host}/websocket/back"
                ws = websocket.create_connection(ws_url)
                try:
                    # 接続時にクライアントIDを送信
                    ws.send(self.client_id)
                    # メッセージを受信するまで待機
                    self.msg = ws.recv()
                    self.got_msg = True
                    return self.msg
                finally:
                    ws.close()
            except (websocket.WebSocketConnectionClosedException, ConnectionRefusedError, ConnectionResetError) as e:
                # メッセージ受信前に接続が閉じた場合、リトライする
                sys.stderr.write(f"WS connection issue: {e}. Retrying after 10s\n")
                time.sleep(10)
            except Exception as e:
                sys.stderr.write(f"An unexpected error occurred in WSIN: {e}\n")
                time.sleep(10)
        return self.msg

class FakeSTDStream:
    """
    sys.stdoutやsys.stderrを置き換えるためのファイル風オブジェクト。
    書き込まれた内容は、元のストリームとリモートエンドポイントの両方に送られます。
    """
    def __init__(self, original_stream, is_err=False):
        self.original_stream = original_stream
        self.is_err = is_err

    def write(self, msg):
        """
        メッセージを元のストリームとリモートの両方に書き込みます。
        """
        # 元のコンソールに書き込む
        self.original_stream.write(msg)
        self.original_stream.flush()
        # リモートエンドポイントに書き込む
        RemoteSTDIO.write(msg, err=self.is_err)

    def flush(self):
        """元のストリームをフラッシュします。"""
        self.original_stream.flush()

    def __getattr__(self, name):
        """
        writeとflush以外のメソッド・属性は、元のストリームオブジェクトに委譲します。
        """
        return getattr(self.original_stream, name)

class RemoteSTDIO:
    """
    標準入出力をリモートで処理するための主要なロジックを管理するクラス。
    RubyのRemoteSTDIOモジュールに相当します。
    """
    # --- クラスレベルの変数で状態を管理 ---
    _host = None
    _client_id = None
    _url = None
    _http_session = None
    _lock = threading.Lock()

    # --- オリジナルの組み込み関数とストリームを保存 ---
    _original_input = builtins.input
    # _original_input = input
    _original_stdout = sys.stdout
    _original_stderr = sys.stderr
    
    _is_initialized = False
    _is_patched = False

    @classmethod
    def init(cls, host, client_id):
        """
        リモート接続を初期化し、標準入出力をパッチします。
        """
        if cls._is_initialized:
            return
            
        cls._host = host
        cls._client_id = client_id
        
        # HTTPセッションの準備
        url_obj = urlparse(f"http://{host}/stdout")
        cls._url = url_obj.geturl()
        cls._http_session = requests.Session()
        
        cls._is_initialized = True
        cls._patch_io()

    @classmethod
    def write(cls, msg, err=False):
        """
        メッセージをJSONペイロードとしてリモートのHTTPサーバーにPOSTします。
        """

        if not msg  or msg == "\n":
            return

        if not cls._is_initialized:
            return

        payload = {
            'message': msg,
            'client_id': cls._client_id,
            'err': err
        }
        try:
            cls._http_session.post(cls._url, json=payload, timeout=5)
        except requests.exceptions.RequestException as e:
            # 元のstderrに書き出すことで無限ループを回避
            cls._original_stderr.write(f"Failed to send message to remote host: {e}\n")

    @classmethod
    def gets(cls, prompt=""):
        """
        ローカルの標準入力とリモートのWebSocketからの入力を同時に待ち受け、
        先に入力があった方を返します。
        """
        if not cls._is_initialized:
            return cls._original_input(prompt)
        
        # プロンプトがあれば先に出力しておく
        if prompt:
            sys.stdout.write(prompt)
            sys.stdout.flush()

        result = [None]  # スレッド間で共有するためのミュータブルなオブジェクト
        got_input_event = threading.Event()

        def read_local_input():
            """ローカルのコンソールからの入力を読み取るスレッド関数。"""
            try:
                local_val = cls._original_input()
                if not got_input_event.is_set():
                    with cls._lock:
                        if not got_input_event.is_set():
                            result[0] = local_val
                            got_input_event.set()
            except EOFError:
                # 入力が終了した場合（Ctrl+Dなど）
                pass
        
        def read_remote_input():
            """リモートのWebSocketからの入力を読み取るスread_remote_inputッド関数。"""
            wsin = WSIN(cls._host, cls._client_id)
            remote_val = wsin.gets()
            if not got_input_event.is_set():
                with cls._lock:
                    if not got_input_event.is_set():
                        result[0] = remote_val
                        got_input_event.set()
        
        # 2つの入力ソースを監視するスレッドを開始
        local_thread = WhineThread(target=read_local_input, daemon=True)
        remote_thread = WhineThread(target=read_remote_input, daemon=True)
        
        local_thread.start()
        remote_thread.start()

        # どちらかのスレッドが完了するのを待つ
        got_input_event.wait()
        
        # 結果を返す
        return result[0] or ''

    @classmethod
    def _patch_io(cls):
        """
        組み込みのinput関数とsys.stdout/stderrを置き換えます。
        """
        if cls._is_patched:
            return
            
        builtins.input = cls.gets
        sys.stdout = FakeSTDStream(cls._original_stdout, is_err=False)
        sys.stderr = FakeSTDStream(cls._original_stderr, is_err=True)
        
        cls._is_patched = True

class RemoteSTDIOUtils:
    """
    環境変数から初期化を行うためのユーティリティクラス。
    """
    @staticmethod
    def init_by_envvar():
        """
        環境変数 'HOST' と 'CID' を使ってRemoteSTDIOを初期化します。
        """
        host = os.getenv('HOST')
        client_id = os.getenv('CID')
        
        if host and client_id:
            RemoteSTDIO.init(host, client_id)
        else:
            # 元のstderrを使用
            RemoteSTDIO._original_stderr.write("WARN: 'HOST' or 'CID' env var not set. Using standard I/O.\n")


# --- メインの実行ブロック ---
if __name__ == "__main__":
    from datetime import datetime

    # 環境変数からHOSTとCIDを取得して初期化
    # 例: export HOST=localhost:3000 CID=my-client-123
    RemoteSTDIOUtils.init_by_envvar()

    # `print` は `sys.stdout` を経由してリモートに送信される
    print("## Hello from Python")
    print(f"**At {datetime.now().isoformat()}**")
    print("~~Test text~~")
    
    # `input` はローカルとリモートの両方で待機する
    # プロンプトは `gets` メソッド内で処理される
    val = input("Input >> ")

    # `sys.stderr` への書き込みもリモートに送信される
    sys.stderr.write(f"Got '{val}'\n")

    print(f"The received value is: {val!r}")